### PR TITLE
Pythnet Oracles

### PR DESCRIPTION
**Pull request by jkbpvsc**
- Add support for pythnet crosschain oracles

TODO:
- Write tests
- Write cranker

### CODE REVIEW

File: programs/marginfi/src/state/native_oracle/mod.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+use self::pyth_crosschain::PythnetPriceFeed;
+
+pub mod pyth_crosschain;
+
+#[repr(C, align(8))]
```
Comment (cavemanloverboy) [inline programs/marginfi/src/state/native_oracle/mod.rs:5]: if using `align(8)` might as well make this repr(u64) as well to ensure the first 8 bytes are properly initialized

File: programs/marginfi/src/state/native_oracle/pyth_crosschain.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            self.price_feed_id.eq(feed_id),
+            MarginfiError::IllegalOracleUpdate,
+            "Price feed ID mismatch"
+        );
+
+        Ok(())
+    }
+
+    /// Check that the verification level is sufficient for the oracle update
+    pub fn check_sufficient_verificaiton_level(
```
Comment (cavemanloverboy) [inline programs/marginfi/src/state/native_oracle/pyth_crosschain.rs:34]: typo: verification

Comment (cavemanloverboy) [reply programs/marginfi/src/state/native_oracle/pyth_crosschain.rs:34]:
```suggestion
    pub fn check_sufficient_verification_level(
```

File: programs/marginfi/src/state/native_oracle/pyth_crosschain.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        &self,
+        verification_level: VerificationLevel,
+    ) -> MarginfiResult<()> {
+        check!(
+            verification_level.gte(self.min_verificaiton_level),
```
Comment (cavemanloverboy) [inline programs/marginfi/src/state/native_oracle/pyth_crosschain.rs:39]: typo here again
```suggestion
            verification_level.gte(self.min_verification_level),
```

File: programs/marginfi/src/state/native_oracle/pyth_crosschain.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            self.publish_time.saturating_add(max_age as i64) >= current_time,
+            MarginfiError::StaleOracle,
+            "Oracle is stale"
+        );
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn get_ema_price(&self) -> MarginfiResult<I80F48> {
```
Comment (cavemanloverboy) [inline programs/marginfi/src/state/native_oracle/pyth_crosschain.rs:162]: may want to add unweighted prefix to be consistent with `get_unweighted_price`:


```suggestion
    fn get_unweighted_ema_price(&self) -> MarginfiResult<I80F48> {
```

File: programs/marginfi/src/constants.rs [hunk 5]
```diff
@@ -120,3 +121,5 @@ pub const EXP_10: [i128; MAX_EXP_10] = [
 
 /// Value where total_asset_value_init_limit is considered inactive
 pub const TOTAL_ASSET_VALUE_INIT_LIMIT_INACTIVE: u64 = 0;
+
+pub const DEFAULT_PYTHNET_VERIFICAITON_LEVEL: VerificationLevel = VerificationLevel::Full;
```
Comment (cavemanloverboy) [inline programs/marginfi/src/constants.rs:125]: typo

File: programs/marginfi/src/instructions/marginfi_group/update_native_oracle_pythnet.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    );
+
+    let pythnet_verificaiton_ctl = bank.get_pythnet_verificaiton_ctl()?;
+
+    match bank.native_oracle {
+        NativeOracle::Pythnet(ref mut native_oracle) => native_oracle.try_update(
+            ctx.accounts.price_update_v2.as_ref(),
+            pythnet_verificaiton_ctl,
+        )?,
+        _ => unreachable!(),
```
Comment (cavemanloverboy) [inline programs/marginfi/src/instructions/marginfi_group/update_native_oracle_pythnet.rs:29]: shouldn't this return error instead?

File: programs/marginfi/src/instructions/marginfi_group/update_native_oracle_pythnet.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+) -> MarginfiResult {
+    let mut bank = ctx.accounts.bank.load_mut()?;
+
+    check!(
+        matches!(bank.config.oracle_setup, OracleSetup::NativePythnet),
+        MarginfiError::InvalidOracleSetup,
+        "Oracle setup is not NativePythnet"
+    );
+
+    let pythnet_verificaiton_ctl = bank.get_pythnet_verificaiton_ctl()?;
```
Comment (cavemanloverboy) [inline programs/marginfi/src/instructions/marginfi_group/update_native_oracle_pythnet.rs:22]: typo

File: programs/marginfi/src/instructions/marginfi_group/update_native_oracle_pythnet.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        ema_price: ctx.accounts.price_update_v2.price_message.ema_price,
+        ema_conf: ctx.accounts.price_update_v2.price_message.ema_conf,
+    });
+
+    Ok(())
+}
+
+#[derive(Accounts)]
+pub struct BankUpdateNativeOraclePythnet<'info> {
+    pub marginfi_group: AccountLoader<'info, MarginfiGroup>,
```
Comment (cavemanloverboy) [inline programs/marginfi/src/instructions/marginfi_group/update_native_oracle_pythnet.rs:52]: why is the group necessary for this? why can't I update native oracle price without the group account? You can read group key from bank when emitting event

**Comment by cavemanloverboy** [comment]
Actually, this doesn't have a unit test that updates the oracle, right?

**Comment by cavemanloverboy** [comment]
i was on a walk and realized this draft PR has an issue. update instruction doesn't check that the price update account is owned by a particular program. it just checks the feed_id. needs to be owned by the receiver program.

**Comment by jkbpvsc** [comment]
Closed for #189

**Pull request closed without merging**